video/out/vo_drm.c

```diff
@@ -0,0 +1,509 @@
+};
+
+struct priv {
+    int fd;
+    struct modeset_dev *dev;
+    drmModeCrtc *old_crtc;
+
+    char *device_path;
+    int connector_id;
+
+    int32_t device_w;
+    int32_t device_h;
+    int32_t x, y;
+    struct mp_image *curframe;
+    struct mp_rect src;
+    struct mp_rect dst;
+    struct mp_osd_res osd;
+    struct mp_sws_context *sws;
+};
+
+static int modeset_open(struct vo *vo, int *out, const char *node)
+{
+    int fd;
+    uint64_t has_dumb;
+    fd = open(node, O_RDWR | O_CLOEXEC);
+    if (fd < 0) {
+        MP_ERR(vo, "Cannot open \"%s\": %s.\n", node, mp_strerror(errno));
+        return -errno;
+    }
+    if (drmGetCap(fd, DRM_CAP_DUMB_BUFFER, &has_dumb) < 0) {
+        MP_ERR(vo, "Device \"%s\" does not support dumb buffers.\n", node);
+        return -EOPNOTSUPP;
+    }
+    *out = fd;
+    return 0;
+}
+
+static int modeset_create_fb(struct vo *vo, int fd, struct modeset_buf *buf)
+{
+    struct drm_mode_create_dumb creq;
+    struct drm_mode_destroy_dumb dreq;
+    struct drm_mode_map_dumb mreq;
+    int ret;
+
+    // create dumb buffer
+    memset(&creq, 0, sizeof(creq));
+    creq.width = buf->width;
+    creq.height = buf->height;
+    creq.bpp = 32;
+    ret = drmIoctl(fd, DRM_IOCTL_MODE_CREATE_DUMB, &creq);
+    if (ret < 0) {
+        MP_ERR(vo, "Cannot create dumb buffer: %s\n", mp_strerror(errno));
+        return -errno;
+    }
+    buf->stride = creq.pitch;
+    buf->size = creq.size;
+    buf->handle = creq.handle;
+
+    // create framebuffer object for the dumb-buffer
+    ret = drmModeAddFB(fd, buf->width, buf->height, 24, 32, buf->stride,
+                       buf->handle, &buf->fb);
+    if (ret) {
+        MP_ERR(vo, "Cannot create framebuffer: %s\n", mp_strerror(errno));
+        ret = -errno;
+        goto err_destroy;
+    }
+
+    // prepare buffer for memory mapping
+    memset(&mreq, 0, sizeof(mreq));
+    mreq.handle = buf->handle;
+    ret = drmIoctl(fd, DRM_IOCTL_MODE_MAP_DUMB, &mreq);
+    if (ret) {
+        MP_ERR(vo, "Cannot map dumb buffer: %s\n", mp_strerror(errno));
+        ret = -errno;
+        goto err_fb;
+    }
+
+    // perform actual memory mapping
+    buf->map = mmap(0, buf->size, PROT_READ | PROT_WRITE, MAP_SHARED,
+                    fd, mreq.offset);
+    if (buf->map == MAP_FAILED) {
+        MP_ERR(vo, "Cannot map dumb buffer: %s\n", mp_strerror(errno));
+        ret = -errno;
+        goto err_fb;
+    }
+
+    memset(buf->map, 0, buf->size);
+
+    return 0;
+
+err_fb:
+    drmModeRmFB(fd, buf->fb);
+err_destroy:
+    memset(&dreq, 0, sizeof(dreq));
+    dreq.handle = buf->handle;
+    drmIoctl(fd, DRM_IOCTL_MODE_DESTROY_DUMB, &dreq);
+    return ret;
+}
+
+static int modeset_find_crtc(struct vo *vo, int fd, drmModeRes *res,
+                             drmModeConnector *conn, struct modeset_dev *dev)
+{
+    drmModeEncoder *enc;
+    unsigned int i, j;
+
+    for (i = 0; i < conn->count_encoders; ++i) {
+        enc = drmModeGetEncoder(fd, conn->encoders[i]);
```

Comment:
I'd prefer if variables are declared on use. (Most of the mpv code does this by now. C89 doesn't allow this, but mpv requires C99.)